Allow multiple db implementations

cammellos opened the pull request:
This commit creates an interface to use with the db so that we can
abstract what kind of db we use, therefore allowing us to chose db based
on config.

It also adds support for postgres, which can be enabled in the config by specifying:

```
{
WhisperConfig:
{ DatabaseConfig:
{PGConfig: { Enabled: true, URI: "postgres://username:password@ip-address:port"}
}
}
```

One host is already running this code:

`mail-01.do-ams3.eth.test`

The DB is still filling up (started listening for envelopes on 2019/05/14 around 1 PM UTC), but it can already be used for testing (performance are not indicative as it's not a full database).

status: ready

cammellos committed:
Allow multiple db implementations

This commit creates an interface to use with the db so that we can
abstract what kind of db we use, therefore allowing us to chose db based
on config.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (30)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 7d0f41da | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/1/) | 2019-05-10 10:30:43 | ~48 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190510-102954-7d0f41.zip) |
| :heavy_check_mark: | 7d0f41da | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/1/) | 2019-05-10 10:33:41 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190510-102953-7d0f41.zip) |
| :heavy_check_mark: | 7d0f41da | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/1/) | 2019-05-10 10:35:44 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190510-102952-7d0f41.aar) |
| | | | | | | |
| :heavy_check_mark: | d21767cf | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/2/) | 2019-05-10 10:54:00 | ~25 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190510-105334-d21767.zip) |
| :heavy_check_mark: | d21767cf | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/2/) | 2019-05-10 10:57:07 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190510-105333-d21767.zip) |
| :heavy_check_mark: | d21767cf | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/2/) | 2019-05-10 10:58:24 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190510-105331-d21767.aar) |
| | | | | | | |
| :heavy_check_mark: | af0e5b1b | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/3/) | 2019-05-10 11:10:54 | ~29 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190510-111023-af0e5b.zip) |
| :heavy_check_mark: | af0e5b1b | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/3/) | 2019-05-10 11:13:10 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190510-111021-af0e5b.zip) |
| :heavy_check_mark: | af0e5b1b | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/3/) | 2019-05-10 11:15:07 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190510-111020-af0e5b.aar) |
| | | | | | | |
| :heavy_check_mark: | 44ff69f9 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/4/) | 2019-05-10 12:55:15 | ~1 min | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190510-125406-44ff69.zip) |
| :heavy_check_mark: | 44ff69f9 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/4/) | 2019-05-10 12:57:10 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190510-125405-44ff69.zip) |
| :heavy_check_mark: | 44ff69f9 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/4/) | 2019-05-10 12:58:54 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190510-125402-44ff69.aar) |
| | | | | | | |
| :heavy_check_mark: | a3e83c07 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/5/) | 2019-05-10 13:16:37 | ~27 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190510-131609-a3e83c.zip) |
| :heavy_check_mark: | a3e83c07 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/5/) | 2019-05-10 13:18:58 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190510-131606-a3e83c.zip) |
| :heavy_check_mark: | a3e83c07 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/5/) | 2019-05-10 13:21:01 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190510-131605-a3e83c.aar) |
| | | | | | | |
| :heavy_check_mark: | fcabc33d | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/6/) | 2019-05-14 07:47:24 | ~31 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190514-074652-fcabc3.zip) |
| :heavy_check_mark: | fcabc33d | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/6/) | 2019-05-14 07:49:37 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190514-074651-fcabc3.zip) |
| :heavy_check_mark: | fcabc33d | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/6/) | 2019-05-14 07:51:42 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190514-074650-fcabc3.aar) |
| | | | | | | |
| :heavy_check_mark: | bf32847f | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/7/) | 2019-05-14 07:56:02 | ~25 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190514-075537-bf3284.zip) |
| :heavy_check_mark: | bf32847f | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/7/) | 2019-05-14 08:00:16 | ~4 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190514-075536-bf3284.zip) |
| :heavy_check_mark: | bf32847f | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/7/) | 2019-05-14 08:00:47 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190514-075534-bf3284.aar) |
| | | | | | | |
| :heavy_check_mark: | 23707ee1 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/8/) | 2019-05-14 12:59:07 | ~39 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190514-125827-23707e.zip) |
| :heavy_check_mark: | 23707ee1 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/8/) | 2019-05-14 13:01:14 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190514-125825-23707e.zip) |
| :heavy_check_mark: | 23707ee1 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/8/) | 2019-05-14 13:03:15 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190514-125824-23707e.aar) |
| | | | | | | |
| :heavy_check_mark: | 3477ce23 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/9/) | 2019-05-14 13:01:06 | ~34 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190514-130031-3477ce.zip) |
| :heavy_check_mark: | 3477ce23 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/9/) | 2019-05-14 13:04:24 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190514-130115-3477ce.zip) |
| :heavy_check_mark: | 3477ce23 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/9/) | 2019-05-14 13:08:06 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190514-130316-3477ce.aar) |
| | | | | | | |
| :heavy_check_mark: | ca26f3ee | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/10/) | 2019-05-15 07:06:16 | ~37 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190515-070539-ca26f3.zip) |
| :heavy_check_mark: | ca26f3ee | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/10/) | 2019-05-15 07:08:36 | ~3 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190515-070536-ca26f3.zip) |
| :heavy_check_mark: | ca26f3ee | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/10/) | 2019-05-15 07:10:36 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190515-070534-ca26f3.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 63ecdacf | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/11/) | 2019-05-15 07:12:27 | ~30 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190515-071156-63ecda.zip) |
| :heavy_check_mark: | 63ecdacf | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/11/) | 2019-05-15 07:14:53 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190515-071156-63ecda.zip) |
| :heavy_check_mark: | 63ecdacf | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/11/) | 2019-05-15 07:17:23 | ~5 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190515-071154-63ecda.aar) |
| | | | | | | |
| :heavy_check_mark: | d6f10d23 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1462/12/) | 2019-05-15 08:39:33 | ~36 sec | `linux` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-190515-083856-d6f10d.zip) |
| :heavy_check_mark: | d6f10d23 | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1462/12/) | 2019-05-15 08:41:49 | ~2 min | `ios` | [:package: zip](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-190515-083855-d6f10d.zip) |
| :heavy_check_mark: | d6f10d23 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1462/12/) | 2019-05-15 08:43:45 | ~4 min | `android` | [:package: aar](https://status-go.ams3.digitaloceanspaces.com/status-go-android-190515-083853-d6f10d.aar) |

mandrigin commented on mailserver/leveldb_impl.go:
```diff
@@ -0,0 +1,113 @@
+package mailserver
```
I would call it `mailserver_db_leveldb.go` and `mailserver_db_postgres.go` instead.

cammellos committed:
Add postgres

This commits adds support for postgres database.
Currently two fields are stored: the bloom filter and the topic.
Only the bloom filter is actually used to query, but potentially we will
use also the topic in the future, so easier to separate it now in order
to avoid a migration.

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

jakubgs commented:
Is the schema created on first connection, or do I need to do some extra step?

Also, please update `config/README.md`.

cammellos commented:
@jakubgs I will update the README, thanks
the scheme is created on first connection, but I need the database set up for me (just an empty db + username & password)

jakubgs commented:
Any specific Postgres version you are expecting?

cammellos commented:
@jakubgs should work fine with any > 8.1

jakubgs commented:
Why the requirement for SSL? If the DB is local it's kinda pointless.
```
error="failed to register MailServer: open DB: pq: SSL is not enabled on the server"
```

cammellos commented:
?sslmode=disable in pg uri

On Tue, May 14, 2019, 13:12 Jakub <notifications@github.com> wrote:

> Why the requirement for SSL? If the DB is local it's kinda pointless.
>
> error="failed to register MailServer: open DB: pq: SSL is not enabled on the server"
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-go/pull/1462?email_source=notifications&email_token=AAHYJMDTESHCFRSD26KBRE3PVKNBFA5CNFSM4HMB3X4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVLEQPQ#issuecomment-492193854>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMD2GTLC55HPMP6B4KLPVKNBFANCNFSM4HMB3X4A>
> .
>

jakubgs commented:
I've deployed `status-go` with PostgreSQL on `mail-01.do-ams3.eth.test`:
https://github.com/status-im/infra-eth-cluster/commit/1fbd6cf36e848dbdb88c665720b811cb0690843f

The connection worked, but I got no log message saying that it did. It might be a good idea to add an `INFO` level message saying it connected to the PostgreSQL database.
And I can see the schema has been created:
```
admin@mail-01.do-ams3.eth.test:~ % d exec -it statusd-mail-db psql -U whisper whisper
psql (9.6.13)
Type "help" for help.

whisper=# \dt
              List of relations
 Schema |       Name        | Type  |  Owner  
--------+-------------------+-------+---------
 public | envelopes         | table | whisper
 public | schema_migrations | table | whisper
(2 rows)
```

jakubgs commented:
Are you fine with just one host? Tell me if you want me to deploy this to the whole `eth.test` cluster.

cammellos commented:
just one host is fine for now, much thank yous! if i need to update the
code, it s the same process as updating the test fleet?

On Tue, May 14, 2019, 13:34 Jakub <notifications@github.com> wrote:

> Are you find with just one host? Tell me if you want me to deploy this to
> the whole eth.test cluster.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-go/pull/1462?email_source=notifications&email_token=AAHYJMAKVJG35WXMERA7QDLPVKPSZA5CNFSM4HMB3X4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVLGCFI#issuecomment-492200213>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMGU2A6XSOC34FY4U5DPVKPSZANCNFSM4HMB3X4A>
> .
>

jakubgs commented:
I've used a specific tag for `mail-01.do-ams3.eth.test`:
https://github.com/status-im/infra-eth-cluster/commit/1fbd6cf36e848dbdb88c665720b811cb0690843f#diff-203160ce66994e01f63e2f353d098f36R3
```
statusd_tag: 'deploy-postgres'
```
So you can update it by doing:
```
make docker-image DOCKER_IMAGE_CUSTOM_TAG=deploy-postgres
docker push statusteam/status-go:deploy-postgres
```

cammellos commented:
goddamn you are efficient, thanks!

On Tue, May 14, 2019, 13:38 Jakub <notifications@github.com> wrote:

> I've used a specific tag for mail-01.do-ams3.eth.test:
> status-im/infra-eth-cluster@1fbd6cf
> #diff-203160ce66994e01f63e2f353d098f36R3
> <https://github.com/status-im/infra-eth-cluster/commit/1fbd6cf36e848dbdb88c665720b811cb0690843f#diff-203160ce66994e01f63e2f353d098f36R3>
>
> statusd_tag: 'deploy-postgres'
>
> So you can update it by doing:
>
> make docker-image DOCKER_IMAGE_CUSTOM_TAG=deploy-postgres
> docker push statusteam/status-go:deploy-postgres
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-go/pull/1462?email_source=notifications&email_token=AAHYJMEMC2TGNMVCIWQBMLLPVKQEXA5CNFSM4HMB3X4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVLGMUI#issuecomment-492201553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYJMAQWKZUYZODALUEO63PVKQEXANCNFSM4HMB3X4A>
> .
>

jakubgs commented:
No problem, poke me if you need anything else.

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

mandrigin commented on mailserver/mailserver.go:
```diff
@@ -111,11 +90,22 @@ func (s *WMailServer) Init(shh *whisper.Whisper, config *params.WhisperConfig) e
-		return fmt.Errorf("open DB: %s", err)
+	if config.DatabaseConfig.PGConfig.Enabled {
+		log.Info("Connecting to postgres database")
+		database, err := NewPostgresImpl(config)
```
I would prefer something like `NewPostgresDB` and `NewLevelDB`, no need to explicitly state that these are implementations, I think it is obvious anyway.

mandrigin commented on mailserver/mailserver_db.go:
```diff
@@ -0,0 +1,36 @@
+
+// dbImpl is an interface to abstract interactions with the db so that the mailserver
+// is agnostic to the underlaying technology used
+type dbImpl interface {
```
I think just `mailserver.db` interface name is okay already. Especially since the interface knows nothing about the implementation, idk if we should have `impl` as a part of the name.

mandrigin commented on mailserver/mailserver_db_leveldb.go:
```diff
@@ -0,0 +1,169 @@
+	"time"
+)
+
+type LevelDBImpl struct {
```
I'd say `type levelDB`: no need to state that that is an implementation and also no need to make it public, especially if the `db` interface isn't.

mandrigin commented on mailserver/mailserver_db_postgres.go:
```diff
@@ -0,0 +1,224 @@
+	return instance, nil
+}
+
+type PostgresImpl struct {
```
`type postgresdb`

mandrigin reviewed (changes requested):
overall the direction is right, but I think it needs a tiny bit more polishing here and there

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty

cammellos commented:
@mandrigin ready to be reviewed, I kept the classes as public because the linter would complain, but otherwise I implemented all the changes.
Thanks!

ghost commented:
`Gopkg.lock` changed. Pinging @adambabik, @mandrigin, and @corpetty
